src/split_at.rs

```diff
@@ -233,59 +213,571 @@ unsafe impl<T> SplitAt for [T] {
     }
 }
 
-/// Splits `T` in two.
-///
-/// # Safety
+/// A `T` that has been split into two possibly-overlapping parts.
 ///
-/// The caller promises that:
-/// 0. `l_len.get()` is not greater than the length of `ptr`'s trailing slice.
-/// 1. if `I::Aliasing` is [`Exclusive`] or `T` permits interior mutation, then
-///    `l_len.padding_needed_for() == 0`.
-#[inline(always)]
-unsafe fn ptr_split_at_unchecked<'a, T, I, R>(
-    ptr: Ptr<'a, T, I>,
-    l_len: MetadataOf<T>,
-) -> (Ptr<'a, T, I>, Ptr<'a, [T::Elem], I>)
+/// For some dynamically sized types, the padding that appears after the
+/// trailing slice field [is a dynamic function of the trailing slice
+/// length](KnownLayout#slice-dst-layout). If `T` is split at a length that
+/// requires trailing padding, the trailing padding of the left part of the
+/// split `T` will overlap the right part. If `T` is a mutable reference or
+/// permits interior mutation, you must ensure that the left and right parts do
+/// not overlap. You can do this at zero-cost using using
+/// [`Self::via_immutable`], [`Self::via_into_bytes`], or
+/// [`Self::via_unaligned`], or with a dynamic check by using
+/// [`Self::via_runtime_check`].
+#[derive(Debug)]
+pub struct Split<T> {
+    /// A pointer to the source slice DST.
+    ///
+    /// # Safety
+    ///
+    /// `source` is a slice DST whose length no less than `l_len`.
+    source: T,
+    /// The length of the future left half of `source`.
+    ///
+    /// # Safety
+    ///
+    /// `l_len` is no greater than `source`'s length.
```

Comment:
IMO you don't need to repeat the safety condition. I'd just put it on `l_len`. It may still make sense to say "`source` is a slice DST."

Reply:
https://github.com/google/zerocopy/compare/708346e8b2599757d96f3e483386d830ed6e8e09..c8a657d1740bfb846580483b3f277a7a3ee58757